src/test/java/htsjdk/samtools/cram/build/ContainerParserTest.java

```diff
@@ -10,177 +10,144 @@
-        Assert.assertEquals(referenceSet.size(), 1);
-        Assert.assertTrue(referenceSet.containsKey(0));
+    @DataProvider(name = "containersForRefTests")
+    private Object[][] refTestData() {
+        return new Object[][] {
+                {
+                        ContainerFactoryTest.getSingleRefRecords(),
+                        new HashSet<Integer>() {{
+                            add(0);
+                        }}
+                },
+                {
+                        ContainerFactoryTest.getUnmappedNoRefRecords(),
+                        new HashSet<Integer>() {{
+                            add(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+                        }}
+                },
+                {
+                        ContainerFactoryTest.getUnmappedNoStartRecords(),
+                        new HashSet<Integer>() {{
+                            add(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+                        }}
+                },
+        };
     }
 
-    @Test
-    public void testUnmappedContainer() throws IOException {
-        SAMFileHeader samFileHeader = new SAMFileHeader();
-        ContainerFactory factory = new ContainerFactory(samFileHeader, 10);
-        List<CramCompressionRecord> records = new ArrayList<>();
-        for (int i=0; i<10; i++) {
-            CramCompressionRecord record = new CramCompressionRecord();
-            record.readBases="AAA".getBytes();
-            record.qualityScores="!!!".getBytes();
-            record.setSegmentUnmapped(true);
-            record.readName=""+i;
-            record.sequenceId= SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX;
-            record.setLastSegment(true);
-
-            records.add(record);
-        }
+    @Test(dataProvider = "containersForRefTests")
+    public void paramTest(final List<CramCompressionRecord> records, final Set<Integer> expectedKeys) {
+        final ContainerFactory factory = new ContainerFactory(ContainerFactoryTest.getSAMFileHeaderForTests(), 10);
+        final Container container = factory.buildContainer(records);
 
-        Container container = factory.buildContainer(records);
-        Assert.assertEquals(container.nofRecords, 10);
-        Assert.assertEquals(container.sequenceId, SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+        final ContainerParser parser = new ContainerParser(ContainerFactoryTest.getSAMFileHeaderForTests());
 
-        ContainerParser parser = new ContainerParser(samFileHeader);
         final Map<Integer, AlignmentSpan> referenceSet = parser.getReferences(container, ValidationStringency.STRICT);
-        Assert.assertNotNull(referenceSet);
-        Assert.assertEquals(referenceSet.size(), 1);
-        Assert.assertTrue(referenceSet.containsKey(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX));
+        Assert.assertEquals(referenceSet.keySet(), expectedKeys);
 
+        // TODO this fails.  return to this when refactoring Container and CramCompressionRecord
+        //
```

Comment:
Any idea why ? For now it might be worth adding an assert that the list sizes match (hopefully....)

Reply:
I think it's due to CramNormalizer?  There's a lot going on there which I don't understand.

More generally, there are many ways to construct a CramCompressionRecord which don't fulfill the requirements of various consumers.  I think some of these are rewriting the records to be more compliant.

Sizes do match, fortunately.